girder/api/sftp.py

```diff
@@ -0,0 +1,302 @@
+        except ValidationException:
+            return paramiko.SFTP_FAILURE
+        except AccessException:
+            return paramiko.SFTP_PERMISSION_DENIED
+        except Exception:
+            logger.exception('SFTP server internal error')
+
+    return wrapped
+
+
+def _getFileSize(file):
+    if file.get('assetstoreId'):
+        return file['size']
+    elif file.get('linkUrl'):
+        return len(file['linkUrl'])
+
+
+class _FileHandle(paramiko.SFTPHandle, ModelImporter):
+    def __init__(self, file):
+        """
+        Create a file-like object representing a file blob stored in Girder.
+
+        :param file: The file object being opened.
+        :type file: dict
+        """
+        super(_FileHandle, self).__init__()
+
+        self.file = file
+
+    def read(self, offset, length):
+        stream = self.model('file').download(
+            self.file, headers=False, offset=offset, endByte=offset + length)
+        return b''.join(stream())
+
+    def stat(self):
+        info = paramiko.SFTPAttributes()
+        info.filename = self.file['name'].encode('utf8')
+        info.st_mtime = time.mktime(self.file['created'].timetuple())
+        info.st_mode = 0o777 | stat.S_IFREG
+        info.st_size = _getFileSize(self.file)
+
+        return info
+
+    def close(self):
+        return paramiko.SFTP_OK
+
+
+class _SftpServerAdapter(paramiko.SFTPServerInterface, ModelImporter):
+    def __init__(self, server, *args, **kwargs):
+        self.server = server
+        paramiko.SFTPServerInterface.__init__(self, server, *args, **kwargs)
+
+    def _list(self, model, document):
+        entries = []
+        if model in ('collection', 'user', 'folder'):
+            for folder in self.model('folder').childFolders(
+                    parent=document, parentType=model, user=self.server.girderUser):
+                info = paramiko.SFTPAttributes()
+                info.st_size = 0
+                info.st_mode = 0o777 | stat.S_IFDIR
+                info.st_mtime = time.mktime(folder['updated'].timetuple())
+                info.filename = folder['name'].encode('utf8')
+                entries.append(info)
+
+        if model == 'folder':
+            for item in self.model('folder').childItems(document):
+                info = paramiko.SFTPAttributes()
+                info.st_size = 0
+                info.st_mode = 0o777 | stat.S_IFDIR
+                info.st_mtime = time.mktime(item['updated'].timetuple())
+                info.filename = item['name'].encode('utf8')
+                entries.append(info)
+        elif model == 'item':
+            for file in self.model('item').childFiles(document):
+                info = paramiko.SFTPAttributes()
+                info.st_mtime = time.mktime(file['created'].timetuple())
+                info.st_mode = 0o777 | stat.S_IFREG
+                info.st_size = _getFileSize(file)
+                info.filename = file['name'].encode('utf8')
+                entries.append(info)
```

Comment:
These blocks seem very similar to the stat method, could we refactor and reuse the same code?